File: backend/dot-net/Services/QuestionService.cs
using backend.Repositories;
using backend.Models;

namespace backend.Services;

public class QuestionService
{
    private readonly QuestionRepository _questionRepository;

    public QuestionService(QuestionRepository questionRepository)
    {
        _questionRepository = questionRepository;
    }

    public Question Create(string title, string description, QuestionStatus status, string userId, string username)
    {
        var question = new Question
        {
            Title = title,
            Description = description,
            Status = status,
            UserId = userId,
            Username = username
        };
        return _questionRepository.Create(question);
    }

    public List<Question> GetAll()
    {
        return _questionRepository.GetAll();
    }

    public (List<Question> Questions, int Total, int Page, int Limit, int TotalPages) GetPaginated(int page, int limit)
    {
        var (questions, total) = _questionRepository.GetPaginated(page, limit);
        return (questions, total, page, limit, (int)Math.Ceiling((double)total / limit));
    }

    public List<Question> Search(string query)
    {
        if (string.IsNullOrWhiteSpace(query))
            throw new ArgumentException("Search query is required");
        return _questionRepository.Search(query);
    }

    public List<Question> GetHot(int limit)
    {
        return _questionRepository.GetHot(limit);
    }

    public Question GetById(string id)
    {
        var question = _questionRepository.GetById(id);
        if (question == null)
            throw new KeyNotFoundException("Question not found");
        return question;
    }

    public List<Question> GetRelated(string id, int limit)
    {
        return _questionRepository.GetRelated(id, limit);
    }

    public Question Update(string id, string title, string description, QuestionStatus status, string userId)
    {
        var existing = _questionRepository.GetById(id);
        if (existing == null)
            throw new KeyNotFoundException("Question not found");
        
        if (existing.UserId != userId)
            throw new UnauthorizedAccessException("You can only edit your own questions");
        
        existing.Title = title;
        existing.Description = description;
        existing.Status = status;
        
        return _questionRepository.Update(existing)!;
    }

    public void Delete(string id, string userId)
    {
        var existing = _questionRepository.GetById(id);
        if (existing == null)
            throw new KeyNotFoundException("Question not found");
        
        if (existing.UserId != userId)
            throw new UnauthorizedAccessException("You can only delete your own questions");
        
        _questionRepository.Delete(id);
    }
}

File: backend/python-fastapi/services/question.service.py
from sqlalchemy.orm import Session
from repositories.question.repository import QuestionRepository


class QuestionService:
    @staticmethod
    def create(db: Session, title: str, description: str, status, user_id: str, username: str):
        return QuestionRepository.create(db, title, description, status, user_id, username)

    @staticmethod
    def get_all(db: Session):
        return QuestionRepository.get_all(db)

    @staticmethod
    def get_paginated(db: Session, page: int, limit: int):
        return QuestionRepository.get_paginated(db, page, limit)

    @staticmethod
    def search(db: Session, query: str):
        if not query:
            raise ValueError("Search query is required")
        return QuestionRepository.search(db, query)

    @staticmethod
    def get_hot(db: Session, limit: int):
        return QuestionRepository.get_hot(db, limit)

    @staticmethod
    def get_by_id(db: Session, id: str):
        question = QuestionRepository.get_by_id(db, id)
        if not question:
            raise ValueError("Question not found")
        return question

    @staticmethod
    def get_related(db: Session, id: str, limit: int):
        return QuestionRepository.get_related(db, id, limit)

    @staticmethod
    def update(db: Session, id: str, title: str, description: str, status, user_id: str):
        existing = QuestionRepository.get_by_id(db, id)
        if not existing:
            raise ValueError("Question not found")
        if existing.user_id != user_id:
            raise ValueError("You can only edit your own questions")
        return QuestionRepository.update(db, id, title, description, status)

    @staticmethod
    def delete(db: Session, id: str, user_id: str):
        existing = QuestionRepository.get_by_id(db, id)
        if not existing:
            raise ValueError("Question not found")
        if existing.user_id != user_id:
            raise ValueError("You can only delete your own questions")
        return QuestionRepository.delete(db, id)
